components/tidb_query_codegen/src/rpn_function.rs

```diff
@@ -519,14 +531,43 @@ impl ValidatorFnGenerator {
 }
 
 fn generate_init_metadata_fn(
-    metadata_ctor: &Option<TokenStream>,
+    metadata_type: &Option<TokenStream>,
+    metadata_mapper: &Option<TokenStream>,
     impl_generics: &ImplGenerics<'_>,
     where_clause: Option<&WhereClause>,
 ) -> TokenStream {
-    let fn_body = if let Some(metadata_ctor) = metadata_ctor {
-        quote! { #metadata_ctor(expr).map(|metadata| Box::new(metadata) as Box<(dyn std::any::Any + std::marker::Send + 'static)>) }
-    } else {
-        quote! { Ok(Box::new(())) }
+    let parse_body_gen = |metadata_type: &TokenStream| {
+        quote! {
+            (if expr.get_val().len() == 0 {
+                Ok(#metadata_type::default())
+            } else {
+                protobuf::parse_from_bytes::<#metadata_type>(expr.get_val())
+                    .map_err(|e| other_err!("Decode metadata failed: {}", e))
+            })
+        }
+    };
+    let fn_body = match (metadata_type, metadata_mapper) {
```

Comment:
I would prefer to extract these logic to `rpn_expr/types/function.rs` so that the code generated from the macro can be fewer, which is easier for debug and read, and maybe more friendly for compiler as well.